flink-runtime/src/main/java/org/apache/flink/runtime/state/SnapshotDirectory.java

```diff
@@ -0,0 +1,172 @@
+
+	/**
+	 * Lifecycle stages of a snapshot directory.
+	 */
+	enum State {
+		ONGOING, COMPLETED, DELETED
+	}
+
+	/** This path describes the underlying directory for the snapshot. */
+	@Nonnull
+	private final Path directory;
+
+	/** The filesystem that contains the snapshot directory. */
+	@Nonnull
+	private final FileSystem fileSystem;
+
+	/** This reference tracks the lifecycle state of the snapshot directory. */
+	@Nonnull
+	private AtomicReference<State> state;
+
+	public SnapshotDirectory(@Nonnull Path directory, @Nonnull FileSystem fileSystem) {
+		this.directory = directory;
+		this.fileSystem = fileSystem;
+		this.state = new AtomicReference<>(State.ONGOING);
+	}
+
+	public SnapshotDirectory(@Nonnull Path directory) throws IOException {
+		this(directory, directory.getFileSystem());
+	}
+
+	@Nonnull
+	public Path getDirectory() {
+		return directory;
+	}
+
+	public boolean mkdirs() throws IOException {
+		return fileSystem.mkdirs(directory);
+	}
+
+	@Nonnull
+	public FileSystem getFileSystem() {
+		return fileSystem;
+	}
+
+	public boolean exists() throws IOException {
+		return fileSystem.exists(directory);
+	}
+
+	/**
+	 * List the statuses of the files/directories in the snapshot directory.
+	 *
+	 * @return the statuses of the files/directories in the given path.
+	 * @throws IOException if there is a problem creating the file statuses.
+	 */
+	public FileStatus[] listStatus() throws IOException {
+		return fileSystem.listStatus(directory);
+	}
+
+	/**
+	 * Calling this method completes the snapshot into the snapshot directory and creates a corresponding
+	 * {@link DirectoryStateHandle} that points to the snapshot directory. Calling this method will also change the
+	 * lifecycle state from "ongoing" to "completed". If the state was already deleted, an {@link IOException} is
+	 * thrown.
+	 *
+	 * @return a directory state handle that points to the snapshot directory.
+	 * @throws IOException if the state of this snapshot directory object is different from "ongoing".
+	 */
+	public DirectoryStateHandle completeSnapshotAndGetHandle() throws IOException {
+		if (state.compareAndSet(State.ONGOING, State.COMPLETED)) {
+			return new DirectoryStateHandle(directory, fileSystem);
+		} else {
+			throw new IOException("Expected state " + State.ONGOING + " but found state " + state.get());
+		}
+	}
+
+	/**
+	 * Calling this method will attempt delete the underlying snapshot directory recursively, if the state was
+	 * "ongoing". In this case, the state will be set to "deleted" as a result of this call.
+	 *
+	 * @return <code>true</code> if delete is successful, <code>false</code> otherwise.
+	 * @throws IOException if an exception happens during the delete.
+	 */
+	public boolean deleteIfIncompleteSnapshot() throws IOException {
+		return state.compareAndSet(State.ONGOING, State.DELETED) && fileSystem.delete(directory, true);
+	}
+
+	public boolean isSnapshotOngoing() {
+		return State.ONGOING.equals(state.get());
```

Comment:
enum values should be compared using `==`.

Reply:
I actually think that `==` is strictly better than `equals` and thus should be used. It has the following advantages: 
- It never throws `NullPointerExceptions`
- It gives you type checks at compile time
- It is faster because `equals` simply defers to `==` internally

Reply:
I think:
- This also never throws `NPE` if you call `equals` on the constant (that is why i emphasized this).
- True.
-Can be optimized to the same thing.

Pro points I see for the other variant:
- does not make you think if `==` is correct in the comparison for even a second.
- still works correctly after refactoring, e.g. when the type is changed from `enum` to `String` constants in case somebody missed to adjust the statement.

Anyways, I will not fight over such a minor thing and can just change it ;)